File: proverb/Proverb.cs
using System;
using System.Linq;

public static class Proverb
{
    public static string[] Recite(string[] subjects) =>
        Enumerable.Range(1, subjects.Length)
            .Select((lineIndex) =>
                (lineIndex == subjects.Length) 
                ? $"And all for the want of a {subjects[0]}."
                : $"For want of a {subjects[lineIndex - 1]} the {subjects[lineIndex]} was lost.")
            .ToArray();
}

File: fsharp/proverb/Proverb.fs
module Proverb

let line (want, lost) =
    sprintf "For want of a %s the %s was lost." want lost

let ending input =
    sprintf "And all for the want of a %s." (input |> List.head)

let recite input =
    match input with
    | [] -> []
    | _ ->
        let lines =
            input
            |> List.pairwise
            |> List.map line
        [ending input] |> List.append lines 